src/main/java/org/jetbrains/bsp/bazel/server/BepServer.java

```diff
@@ -266,132 +234,98 @@ private void processCompletionEvent(BuildEventStreamProtos.BuildEvent event) thr
     }
   }
 
-  private void processActionDiagnostics(BuildEventStreamProtos.BuildEvent event)
-      throws IOException {
-    BuildEventStreamProtos.ActionExecuted action = event.getAction();
-    String actionType = action.getType();
-    if (!SUPPORTED_ACTIONS.contains(actionType)) {
+  private void processAbortedEvent(BuildEventStreamProtos.Aborted aborted) {
+    if (aborted.getReason() != BuildEventStreamProtos.Aborted.AbortReason.NO_BUILD) {
+      buildClientLogger.logError(
+          "Command aborted with reason " + aborted.getReason() + ": " + aborted.getDescription());
+    }
+  }
+
+  private void processActionEvent(BuildEventStreamProtos.ActionExecuted action) throws IOException {
+    if (!Constants.SUPPORTED_ACTIONS.contains(action.getType())) {
       // Ignore file template writes and such.
       // TODO(illicitonion): Maybe include them as task notifications (rather than diagnostics).
       return;
     }
-    System.out.println("DWH: Event: " + event + "\n\n");
-    Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics = new HashMap<>();
+
     BuildTargetIdentifier target = new BuildTargetIdentifier(action.getLabel());
     boolean hasDiagnosticsOutput = diagnosticsProtosLocations.containsKey(target.getUri());
+    Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics = new HashMap<>();
+
     for (BuildEventStreamProtos.File log : action.getActionMetadataLogsList()) {
```

Comment:
it can be done in a functional way too

Reply:
I tried to do it but this code looks suspicious. Why in a loop over _log_ is there something like this (_target_ and _diagnosticsProtosLocations_ are things from outside the loop):

```
if (hasDiagnosticsOutput) {
        diagnosticsProtosLocations.remove(target.getUri());
      }
```

Reply:
This is due to the backwards compatibility with previous bazel versions. We try to cache the location of the diagnostic files based on a query we can do to bazel.

However, if a diagnostic log comes over the BEP, it means that that file a newer version of bazel is being used, therefore, we can delete the cache of the diagnostics file we got beforehand. To do this in a more functional way, you can probably look into filtering for diagnostics with the name `diagnostics` and then if there are any, removing the cache based on the `hasDIagnosticsOutput` flag . Wdyt?

Reply:
Okay, so in the current code

``
if (hasDiagnosticsOutput) {
        diagnosticsProtosLocations.remove(target.getUri());
      }
``

can just be removed from the loop and put at the end of the method?

Reply:
It must only be executed if there's any diagnostics log but yes

Reply:
Wait, then why is it removed anyway after collecting the diagnostics inside ``if (filesToDiagnostics.isEmpty() && hasDiagnosticsOutput)``?

Reply:
If a diagnostics event for a given target has been found but it is empty, then it means that the all of the cached information is now obsolete and that we have no need to go over it with the information from the `diagnosticsProtosLocation`

Reply:
What I mean is, in the original code:

1. If there are some diagnostic logs, then we remove it in the loop.
2. If there aren't any, then filesToDiagnostics is empty, so we remove it anyway at the end.

So I think it can be removed anyway at the end iff ``hasDiagnosticsOutput`` is true.

I've just pushed my version with this change. Could you take a look?

Reply:
Looks good! Good job!